docs/source/debugging.rst

```diff
@@ -0,0 +1,237 @@
+all of your normal Python debugging tricks in Dask computations, as long as you
+don't need parallelism.
+
+This only works for single-machine schedulers.  It does not work with
+dask.distributed unless you use comfortable using the Tornado API (want to look
+at the `testing infrastructure
+<http://distributed.readthedocs.io/en/latest/develop.html#writing-tests>`_
+docs, which accomplish this).  Also, because this operates on a single machine
+it assumes that your computation can run on a single machine without exceeding
+RAM limits.  It may be wise to use this approach on smaller versions of your
+problem if possible.
+
+
+Rerun Failed Task Locally
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If a remote task fails, we can collect the function and all inputs, bring them
+to the local thread, and then rerun the function in hopes of triggering the
+same exception locally, where normal debugging tools can be used.
+
+With the single machine schedulers, use the ``rerun_exceptions_locally=True``
+keyword.
+
+.. code-block:: python
+
+   x.compute(rerun_exceptions_locally=True)
+
+On the distributed scheduler use the ``recreate_error_locally`` method on
+anything that contains ``Futures`` :
+
+.. code-block:: python
+
+   >>> x.compute()
+   ZeroDivisionError(...)
+
+   >>> %pdb
+   >>> future = client.compute(x)
+   >>> client.recreate_error_locally(future)
+
+
+Remove Failed Futures Manually
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes only parts of your computations fail, for example if some rows of a
+CSV dataset are faulty in some way.  When running with the distributed
+scheduler you can remove chunks of your data that have produced bad results if
+you switch to dealing with Futures.
+
+.. code-block:: python
+
+   >>> import dask.dataframe as dd
+   >>> df = ...           # create dataframe
+   >>> df = df.persist()  # start computing on the cluster
+
+   >>> from distributed.client import futures_of
+   >>> futures = futures_of(df)  # get futures behind dataframe
+   >>> futures
+   [<Future: status: finished, type: pd.DataFrame, key: load-1>
+    <Future: status: finished, type: pd.DataFrame, key: load-2>
+    <Future: status: error, key: load-3>
+    <Future: status: pending, key: load-4>
+    <Future: status: error, key: load-5>]
+
+   >>> # wait until computation is done
+   >>> while any(f.status == 'pending' for f in futures):
+   ...     sleep(0.1)
+
+   >>> # pick out only the successful futures and reconstruct the dataframe
+   >>> good_futures = [f for f in futures if f.status == 'finished']
+   >>> df = dd.from_delayed(good_futures, meta=df._meta)
+
+This is a bit of a hack, but often practical when first exploring messy data.
+If you are using the concurrent.futures API (map, submit, gather) then this
+approach is more natural.
+
+
+Inspect Scheduling State
+------------------------
+
+Not all errors present themselves as Exceptions.  For example in a distributed
+systems workers may die unexpectedly or your computation may be unreasonably
```

Comment:
Is it worth mentioning adding `with log_errors():` context management to `distributed/worker.py::apply_function`? Here might not be the most appropriate place though?